src/v/kafka/server/group_metadata.cc

```diff
@@ -0,0 +1,247 @@
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/group_metadata.h"
+
+#include "bytes/bytes.h"
+#include "kafka/protocol/request_reader.h"
+#include "kafka/protocol/response_writer.h"
+#include "kafka/types.h"
+#include "model/adl_serde.h"
+#include "model/fundamental.h"
+#include "model/timestamp.h"
+#include "reflection/adl.h"
+#include "utils/to_string.h"
+#include "vassert.h"
+
+#include <fmt/core.h>
+#include <fmt/ostream.h>
+
+
+namespace kafka {
+
+/**
+ * /Kafka
+ * Messages stored for the group topic has versions for both the key and value
+ * fields. Key version is used to indicate the type of the message (also to
+ * differentiate different types of messages from being compacted together if
+ * they have the same field values); and value version is used to evolve the
+ * messages within their data types:
+ *
+ * key version 0:       group consumption offset
+ *    -> value version 0:       [offset, metadata, timestamp]
+ *
+ * key version 1:       group consumption offset
+ *    -> value version 1:       [offset, metadata, commit_timestamp,
+ * expire_timestamp]
+ *
+ * key version 2:       group metadata
+ *    -> value version 0:       [protocol_type, generation, protocol, leader,
+ * members]
+ */
+group_metadata_type decode_metadata_type(request_reader& key_reader) {
+    auto version = read_metadata_version(key_reader);
+
+    if (
+      version == group_metadata_version{0}
+      || version == group_metadata_version{1}) {
+        return group_metadata_type::offset_commit;
+    }
+
+    if (version == group_metadata_version{2}) {
+        return group_metadata_type::group_metadata;
+    }
+    throw std::invalid_argument(fmt::format(
+      "unexpected group metadata record with key versions {}", version));
+}
+
+void group_metadata_key::encode(
+  response_writer& writer, const group_metadata_key& v) {
+    writer.write(v.version);
+    writer.write(v.group_id);
+}
+
+group_metadata_key
+group_metadata_key::decode(request_reader& reader) {
+    group_metadata_key ret;
+    read_metadata_version(reader);
+    ret.group_id = kafka::group_id(reader.read_string());
+    return ret;
+}
+
+void member_state::encode(
+  response_writer& writer, const member_state& v) {
+    writer.write(v.version);
+    writer.write(v.id);
+    writer.write(v.instance_id);
+    writer.write(v.client_id);
+    writer.write(v.client_host);
+    writer.write(static_cast<int32_t>(v.rebalance_timeout.count()));
+    writer.write(static_cast<int32_t>(v.session_timeout.count()));
+    writer.write(iobuf_to_bytes(v.subscription.copy()));
+    writer.write(iobuf_to_bytes(v.assignment.copy()));
+}
+
+member_state
+member_state::decode(request_reader& reader) {
+    member_state ret;
+    read_metadata_version(reader);
```

Comment:
do you have `vassert` in mind ?

Reply:
I don't have a good sense for what the semantics of the version being read are. So maybe asserting anything about it doesn't make sense. But it's also not clear why a version exists if it can't be used to validate something. Maybe we aren't understanding what metadata version means?